### PR TITLE
feat: Create scope for enqueue in activities

### PR DESCRIPTION
**Pull request by dstenroejl**
## Description

⚠️ Depends on PR for idempotency and should not be merged until that has been merged.

- [x] Fan-out/fan-in
- [x] Perform enumeration within activities
- [x] Create scope for enqueue in activities
- [x] Configure and use retry policies

### CODE REVIEW

File: source/IntegrationTests/Behaviours/IntegrationEvents/TestData/WholesaleResultForAmountPerChargeDescription.cs [hunk 1]
```diff
@@ -47,6 +47,8 @@ public WholesaleResultForAmountPerChargeDescription()
 
     public override string GridAreaCode => "804";
 
+    public override int ExpectedCalculationResultsCount => 7;
+
     public override int ExpectedOutgoingMessagesCount => 14;
```
Comment (RasmusGraabaek) [inline source/IntegrationTests/Behaviours/IntegrationEvents/TestData/WholesaleResultForAmountPerChargeDescription.cs:52]:
```suggestion
    public override int ExpectedOutgoingMessagesCount => 14; // We must prepare a message for ES and ChargeOwner
```

Comment (dstenroejl) [reply source/IntegrationTests/Behaviours/IntegrationEvents/TestData/WholesaleResultForAmountPerChargeDescription.cs:52]: We have refactored this

File: source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs [hunk 2]
```diff
@@ -223,6 +224,41 @@ private static string GetBuildConfiguration()
 #endif
     }
 
+    /// <summary>
+    /// Cleanup Azurite storage to avoid especially Durable Functions
```
Comment (lasrinnil) [inline source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs:228]: why especially?

File: source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueEnergyResultsForBalanceResponsiblesActivity.cs [hunk 3]
```diff
@@ -14,34 +14,55 @@
...
+            input.CalculationId);
+        await foreach (var energyResult in _energyResultEnumerator.GetAsync(query))
         {
-            return 0;
+            using (var scope = _serviceScopeFactory.CreateScope())
+            {
+                try
+                {
+                    var scopedOutgoingMessagesClient = scope.ServiceProvider.GetRequiredService<IOutgoingMessagesClient>();
+                    // TODO: Did not use returned "numberOfEnqueuedMessages", let's talk
```
Comment (lasrinnil) [inline source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueEnergyResultsForBalanceResponsiblesActivity.cs:52]: Could we log how many messages were created? Would be nice to be able to have these numbers quite easy accessible if any discrepancies would come up with how many calculation results were generated and how many messages were

File: source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueWholesaleResultsForAmountPerChargesActivity.cs [hunk 4]
```diff
@@ -14,34 +14,59 @@
...
         {
-            return 0;
+            using (var scope = _serviceScopeFactory.CreateScope())
+            {
+                try
+                {
+                    var scopedMasterDataClient = scope.ServiceProvider.GetRequiredService<IMasterDataClient>();
+                    var scopedOutgoingMessagesClient = scope.ServiceProvider.GetRequiredService<IOutgoingMessagesClient>();
+
+                    // TODO: It should be possible to implement a cache for grid area owner, so we improve the performance of the loop
```
Comment (lasrinnil) [inline source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueWholesaleResultsForAmountPerChargesActivity.cs:53]: lets create a task for this

File: source/BuildingBlocks.Application/Extensions/DependencyInjection/FileStorageExtensions.cs [hunk 5]
```diff
@@ -34,7 +34,7 @@ public static IServiceCollection AddFileStorage(this IServiceCollection services
             .Bind(configuration)
             .Validate(o => !string.IsNullOrEmpty(o.AZURE_STORAGE_ACCOUNT_CONNECTION_STRING) || !string.IsNullOrEmpty(o.AZURE_STORAGE_ACCOUNT_URL), $"{nameof(BlobServiceClientConnectionOptions.AZURE_STORAGE_ACCOUNT_CONNECTION_STRING)} or {nameof(BlobServiceClientConnectionOptions.AZURE_STORAGE_ACCOUNT_URL)} (if using Default Azure Credentials) must be set in configuration");
 
-        services.AddTransient<BlobServiceClient>(
+        services.AddSingleton<BlobServiceClient>(
```
Comment (lasrinnil) [inline source/BuildingBlocks.Application/Extensions/DependencyInjection/FileStorageExtensions.cs:37]: This could be a potential performance increase :D

Comment (dstenroejl) [reply source/BuildingBlocks.Application/Extensions/DependencyInjection/FileStorageExtensions.cs:37]: Its a performance improvement, and the recommendation for any Azure Clients developed by Microsoft. :)
https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-net-clients/

**Review by lasrinnil — DISMISSED** [review]

File: source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueWholesaleResultsForAmountPerChargesActivity.cs (hunk 4 above)
Comment (dstenroejl) [reply source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueWholesaleResultsForAmountPerChargesActivity.cs:53]: Done, will push later

File: source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs (hunk 2 above)
Comment (dstenroejl) [reply source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs:228]: I will just remove the word

File: source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueEnergyResultsForBalanceResponsiblesActivity.cs (hunk 3 above)
Comment (dstenroejl) [reply source/B2BApi/Functions/EnqueueMessages/Activities/EnqueueEnergyResultsForBalanceResponsiblesActivity.cs:52]: I have created tasks on our issue for logging. I will not add it to current PR.

**Comment by dstenroejl** [comment]
⚠️ Depends on PR for idempotency and should not be merged until that has been merged.

File: source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs
Comment (dstenroejl) [inline source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs:1]: ⚠️ Depends on PR for idempotency and should not be merged until that has been merged.

**Review by lasrinnil — DISMISSED** [review]

File: source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs
Comment (ebbeknudsen) [reply source/B2BApi.AppTests/Fixtures/B2BApiAppFixture.cs:1]: Idempotency PR has been merged 👍

**Review by ebbeknudsen — APPROVED** [review]

**Pull request merged**